Initial support for iQue/BBPlayer

rasky opened the pull request:
iQue / BBPlayer is a Chinese variant of Nintendo 64 which has a slightly different hardware and operating system. This PR adds some basic support to libdragon so that it is possible to at least boot the ROMs. ~~The macro N64_BBPLAYER is introduced, and it must be changed to 1 at compile-time to build a iQue-compatible version of libdragon~~. Detection is made at boot time, and at runtime it is possible to call `sys_bbplayer()` to check whether we are running on an iQue Player or not.

The CPU frequency is higher compared to a stock N64, so timers run faster. In this patch, the CPU_FREQUENCY is adjusted so that the timers example pass. Notice that `wait_ticks` is not changed, so games using `wait_ticks` and hardcoding a fixed number of ticks might break with iQue as the delay would be shorter there. We must decide whether to keep this approach, or rather transparently scale all timer-related functions to the original N64 frequency.

A bit is also changed in the display registers otherwise display is garbled. A change is also made to n64tool, to create ROMs which are 16Kb-padded by default, which makes them compatible with iQue. This doesn't affect people explicitly specifying a size (that is still respected), but it is the new default behavior.

This is just the beginning, there are many other differences, but we have to start somewhere.

This PR has been tested on real hardware.

rasky committed:
Initial support for iQue player

CrashOveride95 commented:
> transparently scale all timer-related functions to the original N64 frequency.

This is what the original devs chose to do originally, but bear in mind their priority was for it to work with old code.
Perhaps there's a better way to go about this today
I also remember a address being used to tell if the hardware was a iQue or not, that might be a good idea to avoid having to build seperate libraries for n64/ique, but unsure if this might affect performance

rasky commented:
We had a bit of discussion on Discord and it seems the majority of people feel that it would be vastly better if libdragon supported iQue at runtime, without a compile time switch, so to always generate ROMs which are iQue compatible.

I will need to investigate this if we can find a way without causing too much overhead; I'm concerned about the frequency change mainly, I'll do some tests.

rasky committed:
tests: avoid showing too many DMA failures in test_dma_read_misalign

rasky committed:
n64tool: pad to 16 KiB by default

This is required for a ROM to be loaded on iQue. It shouldn't matter
much for everybody else, and --size is respected anyway when provided,
so this patch only affects the autogenerated size.

Notice that, pending #233, this commit does not introduce a warning
for explicit sizes which are not multiple of 16 KiB. We can add it later
once we have decided the desired semantic for --size.

rasky committed:
README: add mention of iQue support

rasky committed:
iQue: fix dma_read by adjusting the undocumented behavior of PI

dma_read is relying on an undocumented behavior of the hardware to allow
for misaligned transfers. It looks like the PI in iQue is very similar
in its quirks, modulo one single difference (related to short
even-misaligned transfers).

This commit fixes dma_read to behave correctly on iQue too. After this
commit, test_dma_read_misalign pass, and given that it's an exhaustive
bruteforce test, we can be confident that the implementation is sound.

rasky commented:
I have checked the performance. Basically, GCC is able to do what I hoped. For instance, this code:

```C
wait_ticks(TICKS_FROM_MS(50))
```

is expanded as follows:

```C
wait_ticks(50 * ((__bbplayer ? 140625000 : 93750000) / 2 / 1000))
```

which means that it would require at least one division at runtime (depending on the flag). But GCC compiles it to:

```C
__bbplayer ? wait_ticks(3515625) : wait_ticks(2343750);
```

so that the calculations are done at compile-time in the two versions. This is enough to dismiss the concerns I had (even if in this example obviously it doesn't matter if we do one division more if we're going to sleep anyway...).

The new implementation with runtime detection has been tested on real hardware and it works (timers test, which is for now the only one that has been extensively tested). I think we can review and merge this.

sp1187 commented on tools/n64tool.c:
```diff
@@ -428,11 +453,11 @@ int main(int argc, char *argv[])
+	else if(((total_bytes_written+HEADER_SIZE) % PAD_ALIGN) != 0)
 	{
-		/* Pad size to 512 bytes to avoid possible incompatibilities with 64drive
+		/* Pad size to 16384 bytes 512 bytes to avoid possible incompatibilities with 64drive
```
This looks like a typo, the old text should probably be removed.

rasky replied:
Done

rasky commented:
Ok, now the whole test ROM passes on iQue. I've disabled testing the timing of tests as the numbers are different (plus there's something weird going on which I couldn't debug at this moment, there was a floating point exception in timing calculations that didn't make sense -- anyway, maybe it's too hard to debug it in this context). 

I've also adjusted `dma_read` to the slightly-different-and-still-undocumented behaviors of iQue PI DMA for misaligned transfers, so that they work correctly. The audioplayer example now plays beautiful music on both consoles.

sp1187 commented on README.md:
```diff
@@ -15,6 +15,9 @@ programming and debugging. These are the main features:
   to use 64-bit registers and opcodes with it)
 * Can be developed with newer-generation emulators (cen64, Ares, Dillonb's n64,
   m64p) or development cartridges (64drive, EverDrive64).
+* Support both vanilla N64 and iQue Player (chinese variant). The support is
```
Saying that "**all** ROMs will work on both systems" sounds a bit overpromising, it should probably still be considered unfinished and experimental for now. Also, even when it's more mature it will probably still depend on which features are used and if the ROM does anything stupid, buggy or just something out of the ordinary API usage patterns.

rasky replied:
That's true but it's also overly generic. Can we say that all libdragon ROMs work on vanilla N64 either? Not at all, obviously the programmer can always do something that prevents them from working.

I guess it's more a styling issue here. The concept to convey is that you don't need to do anything specific to make ROMs work on iQue. Can you suggest a phrasing that avoids the word "all" which seems to be the one that can be misunderstood?

networkfusion replied:
How about "current libdragon example ROMs and integration tests"

rasky replied:
That's missing the information that we had not to change the examples to make them work. That is, what I would want to say is that the design choice is that user code shouldn't change at all to run on ique

rasky replied:
I've sent a new wording, does this sound better?

sp1187 replied:
Sounds good enough to me.

rasky commented:
Pending #233, I've modified the n64tool commit to avoid the warning for explicit sizes which are not multiple of 16 KiB, which might be surprising for users for now. We're always on time to add it later. The rest is good to go.
